### ISSUE TITLE
Fish-like autocompletion

### ISSUE THREAD

**Issue by felixfbecker** [PowerShell/PSReadLine#687]
I tried to figure this out but couldn't find out. Is there a way to get autocompletion like in the fish shell?

![fish-shell-demo](https://user-images.githubusercontent.com/10532611/40195903-bcb17a06-59c3-11e8-84eb-722775e495e9.gif)

Basically what I am looking for is automatically searching the history on every keystroke, then showing a grey suggestion behind the cursor that can be accepted with <kbd>→</kbd> (if the cursor is at the end of the line).

We can get close with `Set-PSReadLineKeyHandler -Key ArrowRight -Function HistorySearchBackward` but it doesn't show the suggestion and it prevents moving the cursor to the right.

**Comment by lzybkr** [comment]
This feature wasn't much of a priority with limited console support for VT sequences, but now Windows has real support for VT sequences, it makes sense to add this capability.

**Comment by SteveL-MSFT** [comment]
This type of experience is what we envisioned when we originally thought about "Machine Learning autocompletion".  Since then, we dropped the ML part and thought about simply exposing an interface to enable this (https://github.com/PowerShell/PowerShell-RFC/pull/123/files) and just use local history.

**Comment by lzybkr** [comment]
PSReadLine doesn't use `Get-History` - it keeps it's own history.

Here are my thoughts on implementing this feature:

* Add an option to `Set-PSReadLineOption` and document it
* Add a configuration color for the autosuggestion text with a suitable default.
* In `SelfInsert`, you will add smarts to find the autosuggestion. If you want command history, `HistorySearchBackward` has much of what you want, for other completions, `GetCompletions` is probably what you should use.
* The autosuggestion text which is **not** accepted could be stored in a different `StringBuilder`. In `ReallyRender`, you would write this text out with the autosuggestion color.
* In `ForwardChar`, if you are at the end of the buffer **and** there is autosuggestion text, add it to the buffer and clear the autosuggestion.
* Probably add a binding to <kbd>Ctrl+f</kdb> in Windows mode to `ForwardChar`.
* Add a new function `AcceptFirstSuggestedWord` and maybe `AcceptFirstSuggestedShellWord`, binding one or the other to <kbd>Alt+RightArrow</kbd> and <kbd>Alt+f</kbd> if the keys are not already bound (Emacs binds <kbd>Alt+f</kbd> by default, and hmm, maybe <kbd>Alt+RightArrow</kbd> won't work at all.) At any rate, the functions should be available so folks can get this functionality if they want as well.

As for relying on PowerShell for any new fancy suggestions - I'm not excited about that unless it means no api changes. Today a single build of PSReadLine works with any version of PowerShell > 5.0 and I'd like to keep it that way.

**Comment by phiter** [comment]
So, is it coming? :D

**Comment by potatoqualitee** [comment]
Yessss! Please. Looks like there's a whole lot of interest in this. I just saw similar functionality on a livestream and it was awesome!

https://i.vgy.me/PAn45j.mp4

**Comment by daxian-dbw** [comment]
I'm working on a prototype of this, and here is the zip package if any of you want to try it (_any feedback would be very helpful!_).

[PSReadLine.zip](https://github.com/PowerShell/PSReadLine/files/4396121/PSReadLine.zip)

Currently:
- The only suggestion source is the PSReadLine history. The ultimate goal is to call into a prediction API in S.M.A that is powered by a prediction plugin.
- Currently there is no way to configure the style that the suggestion text is rendered with, the default style (color only) works OK with a black background. (of course, more work is needed here)
- VI mode - now the suggestion is disabled when switching to the command mode, and enabled when switching to insert mode.
- So far, I haven't really focused on the VI/Emacs experience, so it could be pretty raw, and feedback is very welcome.

When using this PSReadLine, please add the following key binding to your profile:

    Set-PSReadLineKeyHandler -Key "Ctrl+f" -Function ForwardWord

- Pressing `RightArrow` at the end of current typing will accept the suggestion text
- Pressing `Ctrl+f` at the end of current typing will accept a word from the suggestion text
- Pressing `Ctrl+z` will get you back to where you were before accepting all or part of the suggestion

Important note: don't forget to unblock the zip file after you download it, otherwise you will face weird exceptions thrown from PowerShell.

![predict](https://user-images.githubusercontent.com/127450/74593516-d4391200-4fe0-11ea-9599-3fcbd5836f1e.gif)

**Comment by mpearon** [comment]
This is great to hear!  I would love to see the previous pipeline command considered in the prediction.
For instance,  ```Get-ADUser``` shouldn't suggest ```Get-ACL``` when used in the pipeline.

**Comment by daxian-dbw** [comment]
@mpearon Can you please elaborate on it a bit more? A screenshot would be useful :)

**Comment by daxian-dbw** [comment]
### Issues found

1. prediction shouldn't kick in during history operations like `Ctrl+r`, but it's shown today:
![image](https://user-images.githubusercontent.com/127450/74987986-7007bb00-53f1-11ea-9f54-04461c5aed69.png)

**Comment by mpearon** [comment]
@daxian-dbw This isn't the way it is currently behaving (just a suggestion), so there isn't a screenshot to share.

Maybe I'm missing the point, but my thought was that when ML is leveraged, the suggestions that it makes (not necessarily from history), should take the previous pipeline item into consideration.
So, if I use ``` Get-ADUser ```, the next pipeline commands to be suggested should be those included in that module (or related to the previous command).

**Comment by vexx32** [comment]
Yeah, that probably makes sense. I'd generally assume that you could do something like this, suggested in order of importance or probability to be accurate.

1. Suggest commands that accept the previous command's declared `[OutputType()]` as pipeline input by **value**.
1. Suggest more general utility commands like `Where-Object` and `Select-Object` (and so forth)
1. Suggest commands that accept the previous command's OutputType as pipeline input by **property name**.
1. Suggest other commands from the module that the previous command is in.

I'd also recommend that commands that match either 1 or 3 _and_ are in the same module as the previous command should be suggested first. Any sorting beyond that should probably be related to frequency of usage or just alphabetical perhaps.

**Comment by Southclaws** [comment]
Great to see this being worked on, it's the one feature I miss from Fish!

Question: Have you considered the up/down prefixed history search? So if you type the start of a command then hit the Up Arrow, history searches only through commands with that prefix. Example:


![fish](https://user-images.githubusercontent.com/1636971/75346619-024b0b80-5897-11ea-9cef-dbcb8a2432ce.gif)

**Comment by lzybkr** [comment]
@Southclaws - do you already set these bindings (which I think should be changed to be the default)?

```
Set-PSReadLineKeyHandler -Key UpArrow -Function HistorySearchBackward
Set-PSReadLineKeyHandler -Key DownArrow -Function HistorySearchForward
```

**Comment by Southclaws** [comment]
Oh, amazing, thanks!

Looking forward to the above update too, I'd be interested in testing it out but not sure how to install a module directly from a .zip

**Comment by mpearon** [comment]
@Southclaws - I wasn't entirely sure how to force PS to default to a different PSReadline module either (short of removing/reimporting the module in a profile, or something).
Maybe a bad practice, but I just replaced the contents of the original PSReadline directory. (You can get that by issuing ```(Get-Module PSReadline).path```)

**Comment by mpearon** [comment]
@daxian-dbw Are we creating new issues for the preview here?
After registering the ```Set-PSReadlineKeyHandler``` options mentioned by @lzybkr, my cursor moves to character 0 of the line instead of the end of the command.
I understand that in the case that I've type a bit of a command, this would be desired - but if my cursor is already at character 0, I would expect it to move to the end of the line.
![PSReadLineIssue](https://user-images.githubusercontent.com/7157790/75693449-925ccc80-5c6c-11ea-97ad-fd2719f6eb7e.gif)

**Comment by lzybkr** [comment]
@mpearon - try:
```
Set-PSReadLineOption -HistorySearchCursorMovesToEnd 
```

**Comment by daxian-dbw** [comment]
@mpearon Replacing the original PSReadline directory is the way to use the private zip package. It should be stable enough to not crash. If you decide to get the official PSReadLine module, just use the same "upgrading" instructions from the README.md.

Also, if not yet, take a look at https://github.com/PowerShell/PSReadLine/blob/master/PSReadLine/SamplePSReadLineProfile.ps1. It contains many useful configurations to PSReadLine from @lzybkr.

**Comment by mpearon** [comment]
@lzybkr and @daxian-dbw - Those configuration items have corrected the behavior for me!  Thank you.

**Comment by kort3x** [comment]
just wanted to say: Awesome work - Thank you!

**Comment by SeeminglyScience** [comment]
> * VI mode - now the suggestion is disabled when switching to the command mode, and enabled when switching to insert mode.
> * So far, I haven't really focused on the VI/Emacs experience, so it could be pretty raw, and feedback is very welcome.

Suggestion seems to stay when switching modes, until an edit is made (then it's removed).  I'd like to see it always update even in VI mode personally.  Also `ViForwardChar` doesn't seem to trigger accepting the suggestion.

Very cool though.

**Comment by JustinGrote** [comment]
@daxian-dbw FYI I may be messing something up but I think something changed in the latest 2020.3.1 preview and it stopped working in the integrated console.

EDIT: And yes I made sure to replace it in the extension modules folder

**Comment by daxian-dbw** [comment]
Thanks @JustinGrote for reporting the issue.
Looping in @TylerLeonhardt -- can you please verify it in the latest PS extension preview and see what might be the cause?

**Comment by JustinGrote** [comment]
@TylerLeonhardt Checking the logs it looks like it can't load PSReadline dll for some reason
![image](https://user-images.githubusercontent.com/15258962/77788496-976c2c00-701e-11ea-86af-9c8a7d090af9.png)

If I swap it back with the non-Fish one it works fine.
EDIT: And yes I did check that the files were unblocked since it was a downloaded zip

**Comment by TylerLeonhardt** [comment]
> I think something changed in the latest 2020.3.1 preview

Nothing has change anywhere near the PSReadLine code path so it'd be really bizarre if that update broke it...

> EDIT: And yes I made sure to replace it in the extension modules folder

Run this to double check you're replacing the right one (the one in the extension modules folder is typically for Windows PowerShell):

```
Get-Module PSReadLine | Select-Object Path
```

> Checking the logs it looks like it can't load PSReadline dll for some reason

Very odd... what version of PS? PS7?

**Comment by JustinGrote** [comment]
@TylerLeonhardt if nothing changed in editor services then it's probably me, let me keep banging on it for a bit. For reference it still works in my 5.1 and 7 standalone terminals, just not in Integrated.

**Comment by TylerLeonhardt** [comment]
Looks like we don't support prereleases of PSReadLine:
https://github.com/PowerShell/PowerShellEditorServices/blob/fdb5d7175c8c9c27c639a4d8d7b485b67b8908a6/src/PowerShellEditorServices/Services/PowerShellContext/Session/PSReadLinePromptContext.cs#L30-L46

That was done last month... so the first that showed up was in 2020.2.0 or 2020.3.0 probably...

**Comment by JustinGrote** [comment]
@TylerLeonhardt you nailed it
![image](https://user-images.githubusercontent.com/15258962/77789322-0a29d700-7020-11ea-937e-cc062392a472.png)

The zip has rc2 as a prerelease tag.

EDIT: I remember that I fixed this last time I installed it and then subsequently forgot when the new preview came out yesterday. I'll make a script to make this easier.

**Comment by daxian-dbw** [comment]
I think that check should be removed. There will be newer prerelease coming out for PSReadLine, and that check prevents users from adopting it.

**Comment by TylerLeonhardt** [comment]
Yep :/ if you remove the Prerelease tag from the psd1 it works.

@daxian-dbw we originally had a check that said -ge 2.0.0 & not beta1, beta2, beta3 because a critical PSReadLine fix for PSES was in beta3 or something like that...

```pwsh
Where-Object { $_.PrivateData.PSData.Prerelease -notin 'beta1','beta2','beta3' }
```

Then when PSRL GA'd that check was updated. I agree we shouldn't check for prerelease.

**Comment by SeeminglyScience** [comment]
Maybe it should be:

```powershell
$version -gt '2.0.0' -or ($version -eq '2.0.0' -and -not $_.PrivateData.PSData.Prerelease)
```

**Comment by TylerLeonhardt** [comment]
@SeeminglyScience that would be the ideal since there _are_ bad prereleases in 2.0.0... but then PSRL will need to update to 2.1.0-preview.1 or something for this Fish-like beta... since it's currently still using 2.0.0-rc.2

**Comment by JustinGrote** [comment]
@daxian-dbw can you rebase your autocomplete branch to the latest master? I can have my script pull from your latest appveyor build for us bleeding edge testers.

**Comment by SeeminglyScience** [comment]
> but then PSRL will need to update to 2.1.0-preview.1 or something for this Fish-like beta... since it's currently still using 2.0.0-rc.2

Yeah I'm sure @daxian-dbw plans to increment the version eventually anyway.  Any of us folks running the daily he posted can just alter the manifest for now probably.

**Comment by TylerLeonhardt** [comment]
PSES PR out here: https://github.com/PowerShell/PowerShellEditorServices/pull/1245

**Comment by daxian-dbw** [comment]
@JustinGrote I rebased against `master` and just pushed the [`fish`](https://github.com/daxian-dbw/PSReadLine/tree/fish) branch. But I haven't addressed all the test changes yet, so the CI build in my fork fails and produces no package.

**But, I updated the uploaded .zip package in** https://github.com/PowerShell/PSReadLine/issues/687#issuecomment-586629500, which now uses the 2.0.0 GA version. Hopefully that helps for the time being :)

**Comment by JustinGrote** [comment]
@daxian-dbw  I made an autoinstaller script for the psreadline beta for people to try it out. It installs both for the powershell used to run the script as well as any vscode powershell extensions it finds.

`pwsh.exe -noni -nop -c "iwr -useb https://git.io/InstallPSReadlineAutocompleteBeta.ps1 | iex"`
![Capture](https://user-images.githubusercontent.com/15258962/77939321-da6f0f00-726b-11ea-88d4-abb5d750bd48.gif)


Tested on both win and linux but leave me feedback and comments to issues
https://gist.github.com/JustinGrote/dc10c6120007dceba2da1825ace6b073

**Comment by daxian-dbw** [comment]
Appreciate it @JustinGrote!

**Comment by daxian-dbw** [comment]
To get more feedback, the `v2.1.0-beta1` of PSReadLine was [published](https://github.com/PowerShell/PSReadLine/releases/tag/v2.1.0-beta1), which includes the prototype of the auto-suggestion support. This makes it easy for people to try out.

There is only one addition comparing to the zip package shared before: `PredictionColor` is made configurable through `Set-PSReadLineOption`. So you can choose the color for the suggestion text if the default color doesn't work for you.

**Comment by felixfbecker** [comment]
I just wanted to say I've been using this for multiple weeks and it's AMAZING. Exactly what I imagined.

**Comment by bbhopesh** [comment]
Fish also has smart tab auto-completion for various other commands e.g. `git`. If you write `git add` and hit tab, it will only show files that are changed or untracked. What would be a good way to do the same thing in Powershell? Will it be part of PSReadLine module or some other component?
[https://github.com/fish-shell/fish-shell/blob/master/share/completions/git.fish](https://github.com/fish-shell/fish-shell/blob/master/share/completions/git.fish)

**Comment by TylerLeonhardt** [comment]
@bbhopesh the [`posh-git`](https://github.com/dahlbyk/posh-git) module already supplies this feature:

![image](https://user-images.githubusercontent.com/2644648/79058177-592a5b80-7c1f-11ea-8c96-ca956062dc21.png)

**Comment by felixfbecker** [comment]
The only issue I found is that when you invoke a <kbd>tab</kbd> autocompletion, the history suggestion disappears, even if the new text would still match the history suggestion. Typing one more character then brings it back again. This can feel a bit diruptive/buggy.

**Comment by JustinGrote** [comment]
@daxian-dbw is it possible to wire it up to register-argumentcompleter activities to refresh the history suggestion? after an argumentcompletion is suggested with tab?

**Comment by iSazonov** [comment]
For cross-reference _Predictive IntelliSense_ #1468

**Comment by felixfbecker** [comment]
One thing that stands a bit in the way of these being _as good_ as fish's completion is that in PowerShell, tab-completion completes relative paths with a `./` prefix. This prefix then prevents history suggestions from being matched if you use tab completion and had typed the command before without a `./`, or prevents history suggestions if you submitted only the tab-completed `./` path in the past but are now typing a path without `./`. Generally I don't manually type the `./` because it is redundant (I might as well type `./././` or no prefix, it's the same thing, just more to type).

In fish and bash, paths get completed without a `./` prefix. Is there any reason this is different in PowerShell? Could this perhaps be changed?

**Comment by gilescope** [comment]
Superb! This is excellent. :shipit:

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by vexx32** [comment]
@felixfbecker the de facto reason for that is that if you try to execute a script or other executable in the current location without the `./` or `.\` prefix, PS will refuse to execute the item and prompt you to do that instead.

It might be a bit complicated to make that a little more specific to only apply to `item` and `& item` rather than `application ./item` where it currently is mistakenly applied, but it is probably possible.

**Comment by gilescope** [comment]
Maybe it's just me but I have to specify -NonInteractive when creating powershell in windows terminal for it to do the autosuggest (I'm doing -Command Import-Module from the command line).
- it's _so_ worth it though. It's fantastic.

**Comment by gilescope** [comment]
@felixfbecker it doesn't have to be perfect for v1 - it just needs to be a significant improvement on what came before. That kind of functionallity could be added in later. A fishy MVP :-)

**Comment by Liquidmantis** [comment]
Thanks for the work on this, it's great!  One thing I can't figure out is how to accept the autocomplete when using the vi editmode.  The suggestion is there but neither right arrow nor any vi movements will move to the end in either normal or insert mode.  Do I need to change a keyhandler or something?

**Comment by daxian-dbw** [comment]
@Liquidmantis there are 2 issues in VI mode today:
- Suggestion seems to stay when switching modes
- `ViForwardChar` doesn't seem to trigger accepting the suggestion.

**Comment by chinmayanagpal** [comment]
I just installed the latest release of Powershell Core 7, is there any way to turn this off/replace the suggested text with the tab-complete text instead of history suggestions?

**Comment by JustinGrote** [comment]
@chinmaya1011 that part has not been completed yet, please follow: https://github.com/PowerShell/PSReadLine/issues/1468

And contribute your input for what you would like it to look like if it doesn't meet your needs.

**Comment by mnooseman** [comment]
Hi All - Just found an issue with the autocompletion through an RDP session.  Issue created here: https://github.com/PowerShell/PSReadLine/issues/1558

**Comment by Fooughhy** [comment]
> 
> 
> Maybe it's just me but I have to specify -NonInteractive when creating powershell in windows terminal for it to do the autosuggest (I'm doing -Command Import-Module from the command line).
> 
>     * it's _so_ worth it though. It's fantastic.

I am having an issue with this aswell... I didn't have the "NonInteractive" parameter for my pwsh session before, but I still had this functionality working. Now, all of a sudden, it doesn't work anymore. I have tried to set the "NonInterative" parameter for my pwsh now, but it still doesn't work. I am importing PSReadLine in my profile file (along with oh-my-posh and other things) and it is picking it up when using powershell. But this fish-autocomplete isn't working.

![WindowsTerminal_d1fPclJ1r1](https://user-images.githubusercontent.com/6977416/87310755-ae2b8580-c51e-11ea-8e5e-b491cf8d55a5.png)

**Comment by JustinGrote** [comment]
@Fooughhy can you verify that you are using the new setting? This recently changed to not be on by default, you want to add `set-psreadlineoption -predictionsource history`

`get-psreadlineoption | % predictionsource`

**Comment by Fooughhy** [comment]
There we go! I was trying to find a setting for this, but couldn't... Thank you! It works now.

One more question, does it require "NonInteractive" as my quoted comment suggested? I guess I could try it out.

EDIT: It doesn't seem to be required for me, at least.

**Comment by qubeio** [comment]
This is so awesome - Have all the awards

**Comment by tgeng** [comment]
Thank you for the awesome update! BTW, can function `EndOfLine` accept the suggestion similar to `ForwardChar` as well?

**Comment by lwndwsk1** [comment]
This is epic! Thanks!

**Comment by Praful** [comment]
This is a great new feature. I was looking for it after seeing it on Fish. Thank you for providing it.

I've installed 2.2.0 beta1 and have two questions:

1. What is the equivalent of right-arrow (to accept the prediction) when using Vi mode? Comments above refer to `ViForwardChar` but that doesn't appear to be an option for me. Please can you give an example of setting this to `CTRL-O`.

2. How do I change the prediction colour? Again, an example would be useful. 
**Edit**: answered in #1468
**Edit 2**: I get this error when using `Prediction` as a colour:
```
Set-PSReadLineOption: D:\data\PowerShell\profile.ps1:224
Line |
 224 |  Set-PSReadLineOption -Colors @{ Prediction = [ConsoleColor]::DarkGray …
     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | 'Prediction' is not a valid color property
```
Thanks

**Comment by daxian-dbw** [comment]
@Praful For VI mode, <kbd>RigthArrow</kbd> at the end of the editing line will also accept the prediction:

![demo1](https://user-images.githubusercontent.com/127450/103960237-82244200-5106-11eb-9dba-3fc5091436fc.gif)

You can certainly bind a different key to the `AcceptSuggestion` function as needed.

As for the color, the example in #1468 was out-dated, see the "Change the Color for Predictions" section in https://devblogs.microsoft.com/powershell/announcing-psreadline-2-1-with-predictive-intellisense/

**Comment by Praful** [comment]
@daxian-dbw thank you for answering my question. For those who want to see the answer in one place, I put the following in my PowerShell profile:
```
Set-PSReadLineOption -Colors @{ InlinePrediction = '#000055'}
Set-PSReadLineKeyHandler -chord Ctrl+o -function {ViForwardChar}
```

For those who use Vi mode, they may, like me, not want to take their hands off the home keys to reach for the arrow key, the default key to accept the prediction. I have bound that above to `CTRL+o` for no other reason than it's easy to press from the home row. I used lowercase `O`, simply to avoid having to press `SHIFT` also.

Thanks again for making an essential module even better!

**Comment by yihuajack** [comment]
Thanks for your awesome work. BTW, do you plan to support coloring existing cmdlets, functions, script files, or executable programs with green and non-existing ones with red and underlining paths or filenames users input if they are existing just as ZSH's plugin `zsh-autosuggestions` or fish does? I'm really looking forward to this feature.

**Comment by LuanVSO** [comment]
> Thanks for your awesome work. BTW, do you plan to support coloring existing cmdlets, functions, script files, or executable programs with green and non-existing ones with red and underlining paths or filenames users input if they are existing just as ZSH's plugin `zsh-autosuggestions` or fish does? I'm really looking forward to this feature.

i think those are from `fast-syntax-highlighting`

**Comment by SteveL-MSFT** [comment]
@yihuajack the coloring for existing and non-existing commands seems like a good idea, please open a new issue for this

**Comment by daxian-dbw** [comment]
Given that the original feature request was finished, I will close this issue. Thank you all for all the feedback and comments!
For any new bug reports and enhancement requests, please open new issues for them. Thanks again!

**Comment by potatoqualitee** [comment]
Thanks again for this, it's awesome 🤩

**Comment by JustinGrote** [comment]
> Thanks again for this, it's awesome 🤩

Agreed, use it *constantly*, and looks like Github Copilot aped the same style so it's really consistent looking in my vscode experience these days :)

**Comment by panekj** [comment]
Now we just need it to be *actually* powered by GitHub Copilot! ❤️

**Comment by JustinGrote** [comment]
@panekj you can see the API in use in the console logs in vscode, I'm sure it'd be a flagrant license violation but it would be probably to hijack this session and use the API to fetch Powershell completions :)
![{5F49CE90-A073-47F9-8EED-A74B6D2167C5}](https://user-images.githubusercontent.com/15258962/127000939-e5e0f7bb-5674-4d73-8208-bf431a5b7065.png)

**Comment by hongwen000** [comment]
> Thank you for the awesome update! BTW, can function `EndOfLine` accept the suggestion similar to `ForwardChar` as well?

Found one way to make this possible:
```
Set-PSReadLineKeyHandler -Chord "Ctrl+e" -ScriptBlock {
    [Microsoft.PowerShell.PSConsoleReadLine]::AcceptSuggestion()
    [Microsoft.PowerShell.PSConsoleReadLine]::EndOfLine()
}
```

**Comment by ToxicSmurf** [comment]
Sorry but how do I turn this off?

**Comment by rkeithhill** [comment]
@ToxicSmurf Add this to your profile script to disable predictions:
```powershell
Set-PSReadLineOption -PredictionSource None
```

**Comment by zloyle** [comment]
UpArrow   HistorySearchBackward   Search for the previous item in the history that starts with the current input -
                                  like PreviousHistory if the input is empty
DownArrow HistorySearchForward    Search for the next item in the history that starts with the current input - like
                                  NextHistory if the input is empty
                                  
This is only working for history matches that begin with the string not containing the string the way fish shell supports. Reading the chord binding description in PSReadline, this appears to be intentional.

example:

`$Version = (Get-Module -Name PSReadLine).Version.ToString() `

The only way to find this via HistorySearch is by first typing $ and clicking the UpArrow.

Typing Version or PSReadLine or Module etc. does not work, please look into adding this functionality if time permits.

Regards,
Thank you.

**Comment by owenstake** [comment]
> > Thank you for the awesome update! BTW, can function `EndOfLine` accept the suggestion similar to `ForwardChar` as well?
> 
> Found one way to make this possible:
> 
> ```
> Set-PSReadLineKeyHandler -Chord "Ctrl+e" -ScriptBlock {
>     [Microsoft.PowerShell.PSConsoleReadLine]::AcceptSuggestion()
>     [Microsoft.PowerShell.PSConsoleReadLine]::EndOfLine()
> }
> ```

This method do not work perfect. When cursor is at the begin, "Ctrl+e" leads to complete the prediction.

**Comment by owenstake** [comment]
> > Thank you for the awesome update! BTW, can function `EndOfLine` accept the suggestion similar to `ForwardChar` as well?
> 
> Found one way to make this possible:
> 
> ```
> Set-PSReadLineKeyHandler -Chord "Ctrl+e" -ScriptBlock {
>     [Microsoft.PowerShell.PSConsoleReadLine]::AcceptSuggestion()
>     [Microsoft.PowerShell.PSConsoleReadLine]::EndOfLine()
> }
> ```

This method do not work perfect. When cursor is at the begin, "Ctrl+e" leads to complete the prediction, not to the end of line.

**Comment by owenstake** [comment]
> > Thank you for the awesome update! BTW, can function `EndOfLine` accept the suggestion similar to `ForwardChar` as well?
> 
> Found one way to make this possible:
> 
> ```
> Set-PSReadLineKeyHandler -Chord "Ctrl+e" -ScriptBlock {
>     [Microsoft.PowerShell.PSConsoleReadLine]::AcceptSuggestion()
>     [Microsoft.PowerShell.PSConsoleReadLine]::EndOfLine()
> }
> ```

You method do not work perfect. When cursor is at the begin, "Ctrl+e" leads to complete the prediction, not to the end of line.

I fix it by this sciprt

```powershell
    Set-PSReadLineKeyHandler -Key 'Ctrl+e' `
            -BriefDescription "Go to EndOfLine or AcceptSuggestion" `
            -LongDescription `
                "If cursor is not at the end of line, then go to end of line, `
                otherwise try to AcceptSuggestion" `
            -ScriptBlock {
        $line = $null
        $cursor = $null
        [Microsoft.PowerShell.PSConsoleReadLine]::GetBufferState([ref]$line, [ref]$cursor)
        if ($cursor -ne $line.Length) {
            [Microsoft.PowerShell.PSConsoleReadLine]::EndOfLine()
        } else {
            [Microsoft.PowerShell.PSConsoleReadLine]::AcceptSuggestion()
        }
    }
```

**Comment by yyancy** [comment]
@owenstake Thank you!! It save me a lot of time !!😄
